src/agent/solana/oracle.rs

```diff
@@ -419,88 +417,102 @@ impl Poller {
-                .into_iter()
-                .map(|(product_account, _)| product_account),
-        )
-        .collect();
-
-        let price_accounts = future_results
-            .into_iter()
-            .flat_map(|(_, price_accounts)| price_accounts.into_iter())
-            .collect();
-
-        Ok((product_accounts, price_accounts))
-    }
+        // Look up all products with a single request
+        let product_accounts = self
+            .rpc_client
+            .get_multiple_accounts(product_keys.as_slice())
+            .await?;
 
-    async fn fetch_product_account(
-        &self,
-        product_account_key: &Pubkey,
-    ) -> Result<(ProductAccount, HashMap<Pubkey, PriceAccount>)> {
-        // Fetch the product account
-        let product_account = *load_product_account(
-            &self
-                .rpc_client
-                .get_account_data(product_account_key)
-                .await?,
-        )
-        .with_context(|| format!("load product account {}", product_account_key))?;
-
-        // Fetch the price accounts associated with this product account
-        let mut price_accounts = HashMap::new();
-        let mut price_account_key = product_account.px_acc;
-        while price_account_key != Pubkey::default() {
-            let price_account = self.fetch_price_account(&price_account_key).await?;
-            price_accounts.insert(price_account_key, price_account);
-
-            price_account_key = price_account.next;
+        let mut product_entries = HashMap::new();
+
+        // Log missing products, fill the product entries with initial values
+        for (product_key, product_account) in product_keys.iter().zip(product_accounts) {
+            if let Some(prod_acc) = product_account {
+                let product = load_product_account(prod_acc.data.as_slice())
+                    .context(format!("Could not parse product account {}", product_key))?;
+
+                product_entries.insert(
+                    *product_key,
+                    ProductEntry {
+                        account_data:   *product,
+                        price_accounts: vec![],
+                    },
+                );
+            } else {
+                warn!(self.logger, "Oracle: Could not find product on chain, skipping";
+		      "product_key" => product_key.to_string(),);
+            }
         }
 
-        // Create the product account object
-        let product_account = ProductAccount {
-            account_data:   product_account,
-            price_accounts: price_accounts.keys().cloned().collect(),
-        };
+        let mut price_entries = HashMap::new();
 
-        Ok((product_account, price_accounts))
-    }
+        // Starting with top-level prices, look up price accounts in
+        // batches, filling price entries and adding found prices to
+        // the product entries
+        let mut todo = product_entries
+            .iter()
+            .map(|(_key, p)| p.account_data.px_acc)
```

Comment:
nitpik: based on clippy you can use `product_entires.values().map()`

Reply:
Nice, thanks!